File: src/11.binarySearch/01.374.ts
export function guessNumWrap(n: number, pick: number) {
    function guess(num: number) {
        if (num > pick) {
            return -1;
        } else if (num < pick) {
            return 1;
        }
        return 0;
    }

    function guessNumber(n: number): number {
        let l = 1;
        let r = n;

        while (l < r) {
            const mid = (l + r) >> 1;
            const res = guess(mid);

            if (res === 0) {
                return mid;
            } else if (res === -1) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }

        return r;
    }

    return guessNumber(n);
}

File: src/11.binarySearch/01.374.go
package array_greedy

var pick = 0

func GuessNumberWarp(n, p int) int {
	pick = p
	return GuessNumber(n)
}

func guess(num int) int {
	if num == pick {
		return 0
	}

	if num > pick {
		return -1
	}

	return 1
}

func GuessNumber(n int) int {
	l := 1
	r := n

	for l < r {
		mid := l + (r-l)/2

		res := guess(mid)

		if res == 0 {
			return mid
		}

		if res == -1 {
			r = mid - 1
		} else {
			l = mid + 1
		}
	}

	return r
}
